sdk/core/System.Net.ClientModel/src/Message/MessageResponseHeaders.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Runtime.CompilerServices;
+
+namespace System.Net.ClientModel.Core;
+
+internal class MessageResponseHeaders : MessageHeaders
+{
+    private readonly HttpResponseMessage _httpResponse;
+    private readonly HttpContent _httpResponseContent;
+
+    public MessageResponseHeaders(HttpResponseMessage response, HttpContent responseContent)
+    {
+        _httpResponse = response;
+        _httpResponseContent = responseContent;
+    }
+
+    public override int Count => throw new NotSupportedException();
+
+    public override void Add(string name, string value)
+        => throw new NotSupportedException();
+
+    public override bool Remove(string name)
+        => throw new NotSupportedException();
+
+    public override void Set(string name, string value)
+        => throw new NotSupportedException();
+
+    public override bool TryGetValue(string name, out string? value)
+        => TryGetHeader(_httpResponse.Headers, _httpResponseContent, name, out value);
+
+    public override bool TryGetValues(string name, out IEnumerable<string>? values)
+        => TryGetHeader(_httpResponse.Headers, _httpResponseContent, name, out values);
+
+    public override bool TryGetHeaders(out IEnumerable<KeyValuePair<string, string>> headers)
+    {
+        headers = GetHeadersStringValues(_httpResponse.Headers, _httpResponseContent);
+        return true;
+    }
+
+    public override bool TryGetHeaders(out IEnumerable<KeyValuePair<string, IEnumerable<string>>> headers)
+    {
+        headers = GetHeadersListValues(_httpResponse.Headers, _httpResponseContent);
+        return true;
+    }
+
+    #region Performance-optimized/Platform-specific implementation
+    private static bool TryGetHeader(HttpHeaders headers, HttpContent? content, string name, [NotNullWhen(true)] out string? value)
+    {
+#if NET6_0_OR_GREATER
+        if (headers.NonValidated.TryGetValues(name, out HeaderStringValues values) ||
+            content is not null && content.Headers.NonValidated.TryGetValues(name, out values))
+        {
+            value = JoinHeaderValues(values);
+            return true;
+        }
+#else
+        if (TryGetHeader(headers, content, name, out IEnumerable<string>? values))
+        {
+            value = JoinHeaderValues(values!);
+            return true;
+        }
+#endif
+        value = null;
+        return false;
```

Comment:
Still curious why we need this convenience method that joins the header values for you. But if we need it, I think this could be simplified:
```suggestion
        if (TryGetHeader(headers, content, name, out  IEnumerable<string>? values)
        {
            value = JoinHeaderValues(values);
            return true;
        }        
        value = null;
        return false;
```